physical/raft/raft_autopilot.go

```diff
@@ -0,0 +1,583 @@
+// change. We update a few metrics when this happens.
+func (d *Delegate) NotifyState(state *autopilot.State) {
+	if d.raft.State() == raft.Leader {
+		metrics.SetGauge([]string{"autopilot", "failure_tolerance"}, float32(state.FailureTolerance))
+		if state.Healthy {
+			metrics.SetGauge([]string{"autopilot", "healthy"}, 1)
+		} else {
+			metrics.SetGauge([]string{"autopilot", "healthy"}, 0)
+		}
+
+		for id, state := range state.Servers {
+			if state.Health.Healthy {
+				metrics.SetGauge([]string{"autopilot", string(id), "healthy"}, 1)
+			} else {
+				metrics.SetGauge([]string{"autopilot", string(id), "healthy"}, 0)
+			}
+		}
+	}
+}
+
+// FetchServerStats is called by the autopilot library to retrieve information
+// about all the nodes in the raft cluster.
+func (d *Delegate) FetchServerStats(ctx context.Context, servers map[raft.ServerID]*autopilot.Server) map[raft.ServerID]*autopilot.ServerStats {
+	ret := make(map[raft.ServerID]*autopilot.ServerStats)
+
+	d.l.RLock()
+	followerStates := d.followerStates
+	d.l.RUnlock()
+
+	followerStates.l.RLock()
+	defer followerStates.l.RUnlock()
+
+	now := time.Now()
+	for id, followerState := range followerStates.followers {
+		ret[raft.ServerID(id)] = &autopilot.ServerStats{
+			LastContact: now.Sub(followerState.LastHeartbeat),
+			LastTerm:    followerState.LastTerm,
+			LastIndex:   followerState.AppliedIndex,
+		}
+	}
+
+	leaderState, _ := d.fsm.LatestState()
+	ret[raft.ServerID(d.localID)] = &autopilot.ServerStats{
+		LastTerm:  leaderState.Term,
+		LastIndex: leaderState.Index,
+	}
+
+	return ret
+}
+
+// KnownServers is called by the autopilot library to know the status of each
+// node in the raft cluster. If the application thinks that certain nodes left,
+// it is here that we let the autopilot library know of the same.
+func (d *Delegate) KnownServers() map[raft.ServerID]*autopilot.Server {
+	d.l.RLock()
+	future := d.raft.GetConfiguration()
+	if err := future.Error(); err != nil {
+		d.logger.Error("failed to get raft configuration when computing known servers", "error", err)
+	}
+
+	servers := future.Configuration().Servers
+	serverIDs := make([]string, 0, len(servers))
+	for _, server := range servers {
+		serverIDs = append(serverIDs, string(server.ID))
+	}
+	followerStates := d.followerStates
+	d.l.RUnlock()
+
+	followerStates.l.RLock()
+	defer followerStates.l.RUnlock()
+
+	ret := make(map[raft.ServerID]*autopilot.Server)
+	for id, state := range d.followerStates.followers {
+		// If the server is not in raft configuration, even if we received a follower
+		// heartbeat, it shouldn't be a known server for autopilot.
+		if !strutil.StrListContains(serverIDs, id) {
+			continue
+		}
+
+		server := &autopilot.Server{
+			ID:          raft.ServerID(id),
+			Name:        id,
+			RaftVersion: raft.ProtocolVersionMax,
+			Ext:         d.autopilotServerExt(state.NonVoter),
+		}
+
+		switch state.IsDead {
+		case true:
+			d.logger.Info("informing autopilot that the node left", "id", id)
```

Comment:
This is probably too spammy for Info.

Reply:
This happens only when the node is considered dead by Vault. If cleaning dead servers is enabled, autopilot will knock the node out pretty soon (seconds maybe) and this log will not show up after that. If dead server cleanup is not enabled, this can become spammy, but maybe for good? We will need to inform operators that the node is dead, requiring manual operator intervention to invoke a remove-peer.

Reply:
I believe this will log every 2s by default (autopilot update interval) but we won't do anything about it for a while, since I think the dead server cleanup happens at the reconcile interval (every 10s by default).  Moreover, I believe when autopilot removes the dead server it will log then, making this redundant.

Also consider the case of a user who isn't using autopilot and isn't watching the logs who now risks filling up their log device.  In general I don't like having log messages that can occur at such a high frequency going on indefinitely.  I don't know of any at INFO threshold currently, and I don't think we should start.

Reply:
Sounds good. Reduced to `Debug`.

Reply:
Let me know if you think `Trace` is the right level.

Reply:
So long as it's lower than Info I'm happy.  Thanks!